### PR TITLE
🦙 Azure Storage Add AAD OAuth support

### LINKED ISSUE #164: azure_storage: Add AAD OAuth support
See: https://learn.microsoft.com/en-us/rest/api/storageservices/authorize-with-azure-active-directory

### PR DESCRIPTION
**Pull request by godruoyi**
This PR is to add AAD OAuth support for Azure Storage. Since AAD OAuth essentially authorizes through Access Tokens, we directly added a BearerToken in the Credential.

Please note that we have not implemented how to obtain the Access Token, and this part needs to be implemented by the user themselves.

close #164  ?

maybe we can use https://github.com/Azure/login action to test this changes in the feature.

### CODE REVIEW

File: src/azure/storage/signer.rs [hunk 1]
```diff
@@ -67,6 +67,14 @@ impl Signer {
                 ctx.query_append(token);
                 return Ok(ctx);
             }
+            Credential::BearerToken(token) => {
+                ctx.headers.insert(AUTHORIZATION, {
```
Comment (Xuanwo) [inline src/azure/storage/signer.rs:71]: We can't set header directly without check SigningMethod since we could leak this to user unexpectedly.

**Comment by Xuanwo** [comment]
> close #164 ?

We need to support [the Azure Instance Metadata Service (IMDS)](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http) to close this issue.

**Comment by godruoyi** [comment]
> 

I will implement it in the next PR.

File: src/azure/storage/signer.rs (hunk 1 above)
Comment (godruoyi) [reply src/azure/storage/signer.rs]: 🧑🏾‍💻 update it.

File: src/azure/storage/signer.rs [hunk 2]
```diff
@@ -67,14 +67,17 @@ impl Signer {
...
-            Credential::BearerToken(token) => {
-                ctx.headers.insert(AUTHORIZATION, {
-                    let mut value: HeaderValue = format!("Bearer {}", token).parse()?;
-                    value.set_sensitive(true);
-                    value
-                });
-                return Ok(ctx);
-            }
+            Credential::BearerToken(token) => match method {
+                SigningMethod::Query(_) => {}
```
Comment (Xuanwo) [inline src/azure/storage/signer.rs:71]: We should return an error instead.

**Review by Xuanwo — APPROVED** [review]
Thanks!

**Comment by godruoyi** [comment]
> > We need to support [the Azure Instance Metadata Service (IMDS)](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http) to close this issue.
> 
> I will implement it in the next PR.

hi @Xuanwo, When I was working on this today, I noticed that when requesting an access token for [IMDS](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http), we need to specify the request address as the VM's IP address or use the default 169.254.169.254. However, this default IP now returns "Host is down". If we rely on the VM, it will make our testing difficult. Therefore, I am wondering if we can implement [EnvironmentCredential.get_token](https://github.com/Azure/azure-sdk-for-rust/blob/main/sdk/identity/src/token_credentials/environment_credentials.rs) instead? Can you please help me take a look?

**Comment by Xuanwo** [comment]
> However, this default IP now returns \"Host is down\". If we rely on the VM, it will make our testing difficult. Therefore, I am wondering if we can implement

Yes. This test must be running in Azure Cloud and we can't test it on CI. Sadly, but this is how it works.

**Pull request merged**